core/src/main/c/share/jit/x86.h

```diff
@@ -44,7 +44,14 @@ namespace questdb::x86 {
         c.mov(column_address, ptr(cols_ptr, 8 * column_idx, 8));
         auto shift = type_shift(type);
         auto type_size = 1 << shift;
-        return {Mem(column_address, input_index, shift, 0, type_size), type, data_kind_t::kMemory};
+        if (type_size <= 8) {
+            return {Mem(column_address, input_index, shift, 0, type_size), type, data_kind_t::kMemory};
+        } else {
```

Comment:
can be simplified with:
```
 return {xmmword_ptr(column_address, input_index, shift), type, data_kind_t::kMemory};
```

Reply:
Does casting to `xmmword ptr` here change the x86 rules for calculating scale indexes? If no, I suspect this won't work because the scale factor for the type is too large. As far as I can tell, x86 only allows the scale factor to be 1, 2, 4, or 8; for long128, the type size is 16.

I think what's happening is x86 encodes the scale factor in the instruction as a two-bit shift value; since the shift can thus be 0, 1, 2 or 3, the corresponding allowed scale factors are 1, 2, 4 and 8 respectively. At least, asmjit seems to only use the low two bits of the `scale` parameter being passed here, so if I pass in a `shift` of 4 (`0b100`), asmjit only takes the low `0b00` and the resulting program effectively uses a scale index of 1. At least, that's the perplexing behavior I found my programs had when I first tried JIT-compiling long128 queries :-)

I can give this a try though if you think casting to `xmmword_ptr` will change the way scale factors are handled. Let me know!

Reply:
Had a chance to give this a try today and I believe it doesn't work

Here's a quick repro program

```sql
create table t ( ts timestamp, i int, l1 long128 ) timestamp(ts) partition by day;
< insert values >
select i from t where l1 = to_long128(0, i)
```

If I do the column lookup with `mmword_ptr(column_address, input_index, shift)`, asmjit produces this:

```asm
mov rcx, qword ptr [rdi+8]                  ; mov column_address, qword ptr [cols_ptr+8]
movdqu xmm0, xmmword ptr es:[rcx+rbx]       ; movdqu i128_mem, xmmword ptr es:[column_address+input_index]
```

Note the direct addition of `rcx` (the column base address) and `rbx` (the input index for the loop). This happens because asmjit only looks at the bottom two bits of shift, sees they're zero, and so emits an instruction where shift is zero. The result is a query that correctly reads the memory reference for the first row but reads unaligned junk after.

The thing I did is clunky, but does result in a correct column lookup:

```asm
mov rdx, qword ptr [rdi+8]                  ; mov column_address, qword ptr [cols_ptr+8]
mov rcx, 16                                 ; mov row_offset, 16
imul rcx, rbp                               ; imul row_offset, input_index
movdqu xmm0, xmmword ptr [rdx+rcx]          ; movdqu i128_mem, xmmword ptr [column_address+row_offset]
```

I realize this is an extra register and four instructions where we used to do just one, so I'm all for finding a faster way to do this if you can think of one.